File: frontend/shared/api_client.js
/**
 * JavaScript API client for Voice App backend
 */
class VoiceAPIClient {
  constructor(baseUrl = "http://localhost:8000") {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.defaultHeaders = {
      "Content-Type": "application/json",
      "User-Agent": "VoiceApp-WebClient/1.0",
    };
  }

  /**
   * Make HTTP request with error handling
   */
  async request(endpoint, options = {}) {
    const url = `${this.baseUrl}${endpoint}`;
    const config = {
      headers: { ...this.defaultHeaders, ...options.headers },
      ...options,
    };

    try {
      const response = await fetch(url, config);
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }
      return await response.json();
    } catch (error) {
      console.error(`Request failed: ${error.message}`);
      return { error: error.message };
    }
  }

  /**
   * Check backend health status
   */
  async healthCheck() {
    return await this.request("/healthz");
  }

  /**
   * Convert text to speech
   */
  async textToSpeech(text, options = {}) {
    const payload = {
      text,
      speaker_id: options.speakerId || "default",
      language: options.language || "zh",
      speed: options.speed || 1.0,
    };

    return await this.request("/api/v1/tts", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  }

  /**
   * Convert voice to target speaker
   */
  async voiceConversion(audioFile, targetSpeaker, options = {}) {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();

      reader.onload = async () => {
        const arrayBuffer = reader.result;
        const uint8Array = new Uint8Array(arrayBuffer);
        const audioB64 = this.arrayBufferToBase64(uint8Array);

        const payload = {
          source_audio: audioB64,
          target_speaker: targetSpeaker,
          preserve_pitch: options.preservePitch !== false,
        };

        const result = await this.request("/api/v1/vc", {
          method: "POST",
          body: JSON.stringify(payload),
        });
        resolve(result);
      };

      reader.onerror = () => reject(new Error("File read failed"));
      reader.readAsArrayBuffer(audioFile);
    });
  }

  /**
   * Get available speaker profiles
   */
  async getProfiles() {
    return await this.request("/api/v1/profiles");
  }

  /**
   * Batch text-to-speech conversion
   */
  async batchTTS(texts, options = {}) {
    const payload = {
      texts,
      ...options,
    };

    return await this.request("/api/v1/batch/tts", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  }

  /**
   * Get audio file URL for playback
   */
  getAudioUrl(audioPath) {
    if (audioPath.startsWith("/")) {
      return `${this.baseUrl}${audioPath}`;
    }
    return audioPath;
  }

  /**
   * Download audio file
   */
  async downloadAudio(audioUrl, filename) {
    try {
      const fullUrl = this.getAudioUrl(audioUrl);
      const response = await fetch(fullUrl);

      if (!response.ok) {
        throw new Error(`Download failed: ${response.statusText}`);
      }

      const blob = await response.blob();

      // Create download link
      const url = window.URL.createObjectURL(blob);
      const a = document.createElement("a");
      a.href = url;
      a.download = filename || "audio.wav";
      document.body.appendChild(a);
      a.click();
      document.body.removeChild(a);
      window.URL.revokeObjectURL(url);

      return true;
    } catch (error) {
      console.error("Download failed:", error);
      return false;
    }
  }

  /**
   * Convert ArrayBuffer to Base64
   */
  arrayBufferToBase64(buffer) {
    let binary = "";
    const bytes = new Uint8Array(buffer);
    const len = bytes.byteLength;
    for (let i = 0; i < len; i++) {
      binary += String.fromCharCode(bytes[i]);
    }
    return window.btoa(binary);
  }

  /**
   * Validate audio file type
   */
  isValidAudioFile(file) {
    const validTypes = ["audio/wav", "audio/mp3", "audio/ogg", "audio/m4a"];
    return validTypes.includes(file.type);
  }

  /**
   * Format file size for display
   */
  formatFileSize(bytes) {
    if (bytes === 0) return "0 Bytes";
    const k = 1024;
    const sizes = ["Bytes", "KB", "MB", "GB"];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + " " + sizes[i];
  }
}

// Export for different module systems
if (typeof module !== "undefined" && module.exports) {
  module.exports = VoiceAPIClient;
}

if (typeof window !== "undefined") {
  window.VoiceAPIClient = VoiceAPIClient;
}

File: frontend/shared/api_client.py
# frontend/shared/api_client.py

import requests
import base64
import json
from pathlib import Path
from typing import Dict, Any, Optional, Union
import logging

logger = logging.getLogger(__name__)


class VoiceAPIClient:
    """Python API client for Voice App backend"""

    def __init__(self, base_url: str = "http://localhost:8000"):
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()
        self.session.headers.update(
            {
                "Content-Type": "application/json",
                "User-Agent": "VoiceApp-PythonClient/1.0",
            }
        )

    def health_check(self) -> Dict[str, Any]:
        """Check backend health status"""
        try:
            response = self.session.get(f"{self.base_url}/healthz")
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"Health check failed: {e}")
            return {"status": "error", "message": str(e)}

    def text_to_speech(
        self,
        text: str,
        speaker_id: str = "default",
        language: str = "zh",
        speed: float = 1.0,
    ) -> Dict[str, Any]:
        """Convert text to speech"""
        payload = {
            "text": text,
            "speaker_id": speaker_id,
            "language": language,
            "speed": speed,
        }

        try:
            response = self.session.post(f"{self.base_url}/api/v1/tts", json=payload)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"TTS request failed: {e}")
            return {"error": str(e)}

    def voice_conversion(
        self,
        audio_file: Union[str, Path, bytes],
        target_speaker: str,
        preserve_pitch: bool = True,
    ) -> Dict[str, Any]:
        """Convert voice to target speaker"""

        # Handle different input types
        if isinstance(audio_file, (str, Path)):
            with open(audio_file, "rb") as f:
                audio_data = f.read()
        else:
            audio_data = audio_file

        # Encode to base64
        audio_b64 = base64.b64encode(audio_data).decode("utf-8")

        payload = {
            "source_audio": audio_b64,
            "target_speaker": target_speaker,
            "preserve_pitch": preserve_pitch,
        }

        try:
            response = self.session.post(f"{self.base_url}/api/v1/vc", json=payload)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"VC request failed: {e}")
            return {"error": str(e)}

    def get_profiles(self) -> Dict[str, Any]:
        """Get available speaker profiles"""
        try:
            response = self.session.get(f"{self.base_url}/api/v1/profiles")
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"Get profiles failed: {e}")
            return {"profiles": [], "error": str(e)}

    def batch_tts(self, texts: list, **kwargs) -> Dict[str, Any]:
        """Batch text-to-speech conversion"""
        payload = {"texts": texts, **kwargs}

        try:
            response = self.session.post(
                f"{self.base_url}/api/v1/batch/tts", json=payload
            )
            response.raise_for_status()
            return response.json()
        except requests.exceptions.RequestException as e:
            logger.error(f"Batch TTS failed: {e}")
            return {"error": str(e)}

    def download_audio(self, audio_url: str, save_path: Union[str, Path]) -> bool:
        """Download audio file from backend"""
        try:
            full_url = (
                f"{self.base_url}{audio_url}"
                if audio_url.startswith("/")
                else audio_url
            )
            response = self.session.get(full_url)
            response.raise_for_status()

            with open(save_path, "wb") as f:
                f.write(response.content)
            return True
        except requests.exceptions.RequestException as e:
            logger.error(f"Download failed: {e}")
            return False
